s3-proxy/nginx.conf

```diff
@@ -58,17 +58,29 @@ http {
             # Use virtual hosts by default, but path-style for buckets with dots.
             set $s3_host '$s3_bucket.$s3_host_suffix';
             set $s3_path_prefix '';
+            set $s3_path_sep '';
             if ($s3_bucket ~ "\.") {
                 set $s3_host '$s3_host_suffix';
-                set $s3_path_prefix '$s3_bucket/';
+                set $s3_path_prefix '$s3_bucket';
+                set $s3_path_sep '/';
             }
 
-            # Proxy the request to S3.
+            set $s3_path_suffix '';
             # Use $request_uri rather than $s3_path because it needs to stay encoded.
             if ($request_uri ~ "^/[^/?]+/[^/?]+/?(.*)") {
-                proxy_pass 'https://$s3_host/$s3_path_prefix$1';
+                set $s3_path_suffix "$1";
```

Comment:
I've run into this before... It's likely an Nginx bug.

This will keep URL encoding:
```
  proxy_pass 'https://$s3_host/$s3_path_prefix$s3_path_sep$1;
```

But this will break it:

```
  set $s3_path_suffix "$1";
  proxy_pass 'https://$s3_host/$s3_path_prefix$s3_path_sep$s3_path_suffix';
```

Even though they are supposed to be exactly the same thing.

Will need to work around it somehow. Worst case, have two `if`s, for requests with and without path segments.